### PR TITLE
Validate recipient email addresses to prevent invalid ending characters

### LINKED ISSUE #7112: The recipient address <xxx@gmail.com/> is not a valid RFC-5321 address.
### Checklist

- [X] I have used the search function to see if someone else has already submitted the same bug report.
- [X] I will describe the problem with as much detail as possible.

### App version

v6.602

### Where did you get the app from?

Google Play

### Android version

13

### Device model

Moto X40

### Steps to reproduce

Edit a mail, input the recipient address end with a `/`, then send it.

https://github.com/thundernest/k-9/assets/10363352/cbb7a99a-b348-4f88-adc3-3fccc1f4ba29

### Expected behavior

We can check the mail address before sending it, toast a message to tell the user that the address is invalid, or show a dialog like Gmail did below:

### Actual behavior

Sending it, and showing the failure.

### Logs

_No response_

### PR DESCRIPTION
**Pull request by mhf1998**
Fixes #7112 

- Added a regular expression pattern to identify and invalidate email addresses that end with specific invalid characters.
- The validation logic was incorporated into the `isValidEmail` method, ensuring that any email address ending with characters such as `/`, `?`, `=`, `+`, `&`, `^`, `%`, `$`, `#`, `!`, `'`, and `-` is correctly flagged as invalid. (I checked and all the special characters that led to the repeating of this bug scenario were just these.)
- Improved error handling, and show an error when the focus is changed from the To, CC, and BCC fields; instead of showing the error after sending the email.

image

### CODE REVIEW

**Review by cketti — CHANGES_REQUESTED** [review]
Please add an additional (in)validity check for the domain part to `UserInputEmailAddressParser` and add relevant tests to `UserInputEmailAddressParserTest`.

Side note: This doesn't fix the underlying problem of us (indirectly) using `Rfc822Tokenizer` which is a bad fit for the job. However, it does improve the current situation a little bit.

File: legacy/ui/legacy/src/main/java/com/fsck/k9/view/RecipientSelectView.java [hunk 1]
```diff
@@ -182,13 +184,35 @@ protected Recipient defaultObject(String completionText) {
...
         } catch (NonAsciiEmailAddressException e) {
             setError(getContext().getString(R.string.recipient_error_non_ascii));
             return null;
         }
     }
 
+    // Method to validate email addresses
+    private boolean isValidEmail(List<Address> addresses) {
+        // Regular expression pattern to identify invalid ending characters in email addresses
+        Pattern invalidEndingPattern = Pattern.compile("[/?=+&^%$#!'\\-]$");
```
[the remark this answers is not in the document]
Comment (mhf1998) [reply legacy/ui/legacy/src/main/java/com/fsck/k9/view/RecipientSelectView.java:203]: > A quick test revealed that adding e.g. a slash in the middle of the domain part is currently also considered valid by `UserInputEmailAddressParser`. Example: `test@domain/example`.
> 
> So only checking the last character of the domain part is not sufficient to catch all instances of this bug. I suggest using `HostNameUtils.isLegalHostNameOrIP()` to validate the domain part of the entered email address.

I read everything you mentioned. Got it. 
I'll get it all done at the first chance I get.

File: legacy/ui/legacy/src/main/java/com/fsck/k9/view/RecipientSelectView.java [hunk 2]
```diff
@@ -182,13 +184,35 @@ protected Recipient defaultObject(String completionText) {
                 return null;
             }
 
+            // email address validation logic after parsing
+            if (!isValidEmail(parsedAddresses)) {
```
[the remark this answers is not in the document]
Comment (mhf1998) [reply legacy/ui/legacy/src/main/java/com/fsck/k9/view/RecipientSelectView.java]: Done!

File: legacy/ui/legacy/src/main/java/com/fsck/k9/view/RecipientSelectView.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (mhf1998) [reply legacy/ui/legacy/src/main/java/com/fsck/k9/view/RecipientSelectView.java]: Done!

**Comment by mhf1998** [comment]
I have implemented everything you asked for. Now, the domain is correctly validated by the use of the function you mentioned, and even if the last character of the email is wrong, that is caught as well, and the relevant error is displayed. 
Moreover, all tests have been written and added to the class.

File: legacy/ui/legacy/src/main/res/values-fa/strings.xml [hunk 3]
```diff
@@ -950,4 +950,5 @@
     <string name="dialog_openkeychain_info_text">برنامه کاربردی OpenKeychain جهت فعال‌سازی رمزگذاری سرتاسری مورد نیاز است.</string>
     <string name="push_info_notification_explanation_text">در زمان استفاده از پیشرانی، <xliff:g id="app_name">%1$s</xliff:g> اتصال به کارساز را نگاه می‌دارد. اندروید نیازمند یک اعلان همیشگی جهت فعال ماندن برنامه در پس‌زمینه است. <xliff:g id="optional_extended_explanation">%2$s</xliff:g></string>
     <string name="about_app_authors_thunderbird">گروه همراه تاندربرد</string>
-</resources>
\ No newline at end of file
+    <string name="recipient_error_invalid_email">آدرس ایمیل نامعتبر است</string>
```
Comment (cketti) [inline legacy/ui/legacy/src/main/res/values-fa/strings.xml:953]: Translations should only be updated via Weblate. Changing them in a pull request leads to the potential for merge conflicts in Weblate that are annoying to fix.

**Comment by mhf1998** [comment]
@cketti
I have implemented everything you asked for. Done.

**Review by cketti — APPROVED** [review]

**Comment by cketti** [comment]
Thanks :+1:

**Comment by mhf1998** [comment]
You're welcome.
It's an honor to collaborate on this project. ☺️
Thank you for taking the time to review and respond.🙏🏻

**Pull request merged**